tests/php/test-class-amp-srcset-sanitizer.php

```diff
@@ -0,0 +1,126 @@
+				null,
+			],
+
+			'img_with_duplicate_img_candidate_but_same_url' => [
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg, https://example.com/image-1.jpg     1024w, https://example.com/image-1.jpg 1024w   , https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg 1x, https://example.com/image-1.jpg 1024w, https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+			],
+
+			'img_with_duplicate_img_candidate_but_different_url' => [
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg, https://example.com/image-1.jpg     1024w, https://example.com/image-2.jpg 1024w   , https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg 1x, https://example.com/image-1.jpg 1024w, https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+				[ AMP_Tag_And_Attribute_Sanitizer::DUPLICATE_DIMENSIONS ],
+			],
+
+			'amp_img_srcset_missing_comma'              => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="https://example.com/image-1.jpg 1024w https://example.com/image-2.jpg 1024w">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+
+			'amp_img_srcset_empty'                      => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="">',
+				null,
+				[],
+			],
+
+			'amp_img_srcset_whitespace'                 => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="    ">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+
+			'amp_img_srcset_invalid_bare_number'        => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="1">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+
+			'amp_img_srcset_invalid_bare_dimension'     => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="https://example.com/image.jpg 500px">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+
+			'amp_img_srcset_invalid_space_in_dimension' => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="https://example.com/image.jpg 500px">',
```

Comment:
Same as previous test.

Reply:
“Fixed” in 8541363c2be982c029efb9d290ba79c0f6eda399, but nevertheless this test now fails:

```
1) AMP_Srcset_Sanitizer_Test::test_sanitize with data set "amp_img_srcset_invalid_space_in_dimension" ('<img src="https://example.com... 2 x">', '<img src="https://example.com..."200">', array('INVALID_ATTR_VALUE'))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
+    0 => 'INVALID_ATTR_VALUE'

/app/public/core-dev/tests/phpunit/includes/abstract-testcase.php:685
/app/public/content/plugins/amp/tests/php/test-class-amp-srcset-sanitizer.php:120
``` 

It's not flagging `https://example.com/image.jpg 2 x` as invalid.

Reply:
In this case `$matches` is:

```json
{
    "0": [
        "https://example.com/image.jpg "
    ],
    "url": [
        "https://example.com/image.jpg"
    ],
    "1": [
        "https://example.com/image.jpg"
    ],
    "dimension": [
        ""
    ],
    "2": [
        ""
    ],
    "comma": [
        ""
    ],
    "3": [
        ""
    ]
}
```

And because `dimension` is just `[ "" ]` then that means this will return incorrectly:

https://github.com/ampproject/amp-wp/blob/8541363c2be982c029efb9d290ba79c0f6eda399/includes/sanitizers/class-amp-srcset-sanitizer.php#L76-L80

@pierlon it looks like either the regex or the `$matches` examination needs to be hardened.

Reply:
I've added a fix that should short-circuit and return if the last full match from the regex pattern does not match the end of the `srcset` string:

https://github.com/ampproject/amp-wp/blob/665c669f48e00f39142321acd37611a5bc5aff4f/includes/sanitizers/class-amp-srcset-sanitizer.php#L58

Reply:
Nicely done.